### PR TITLE
Allow sync loading of ESM when `--experimental-require-module`

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | `Fixes #1, Fixes #2` 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Node.js now has this flag to allow `require()` of ESM files, which means we can support ESM plugins/presets in `transformSync`&co. This will be useful for the Babel 8 migration.

This PR simply makes us respect that Node.js flag rather than just refusing to `require()` esm files ending in `.mjs`. I'm not sure if I should consider it as a bugfix or as internal. Note that `require("<something>.js")` already worked for ESM under that flag, because we try/catch to see if the `.js` file can be loaded.

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
This is not breaking but I'd like to get it merged before the 8.0.0-beta, since it significantly simplify using ESM plugins (i.e. you wouldn't be forced to use Babel asynchronously anymore).

File: packages/babel-core/src/config/files/module-types.ts [hunk 1]
```diff
@@ -79,16 +83,29 @@ const loadMjsDefault = endHiddenCallStack(async function loadMjsDefault(
...
   }
 });
 
+const SUPPORTED_EXTENSIONS = new Set([".js", ".mjs", ".cjs", ".cts"] as const);
+type SetValue<T extends Set<unknown>> = T extends Set<infer U> ? U : never;
+
 export default function* loadCodeDefault(
   filepath: string,
-  asyncError: string,
+  loader: "require" | "import" | "auto",
```
Comment (liuxingbaoyu) [inline packages/babel-core/src/config/files/module-types.ts:95]: `auto``auto` doesn't seem to be used?

**Comment by nicolo-ribaudo** [comment]
This flag will be unflagged in Node.js 23.

**Review by liuxingbaoyu — APPROVED** [review]
Amazing!

**Review by JLHwung — APPROVED** [review]

**Comment by victordidenko** [comment]
I'm not sure, but looks like this change breaks something, I'm receiving error
```
[!] (plugin wyw-in-js) Error: [BABEL] src/index.ts: .loader is not a valid Plugin property
src/index.ts
    at .../node_modules/.pnpm/@babel+core@7.25.7/node_modules/@babel/core/src/config/validation/plugins.ts:122:41
    at Array.forEach (<anonymous>)
    at validatePluginObject (.../node_modules/.pnpm/@babel+core@7.25.7/node_modules/@babel/core/src/config/validation/plugins.ts:108:20)
```
Or maybe this is an issue with https://github.com/Anber/wyw-in-js/issues/104

**Comment by nicolo-ribaudo** [comment]
Could you open a new issue? This is indeed a regression, I didn't realize that that API is public.

**Comment by victordidenko** [comment]
[opened](https://github.com/babel/babel/issues/16887) 👌

**Pull request merged**